### DISCUSSION TITLE
How to use global mixins in nuxt 3

### DISCUSSION THREAD

**Discussion by Trevash** [nuxt/nuxt#16706] [Questions] [answered]
Hey I'm trying to import and add a global mixin file to be used on all pages, but the technique I used in Nuxt 2 doesn't seem to be available in Nuxt 3. The reason I want to add a global mixin file is so that I have access to functions I'll use on almost every page without having to import the mixin each time.

Here is what I have done in the past in either my app.vue file or my layouts/default.vue file depending on my file structure:
```
import Vue from 'vue'
import globalMixin from '~/mixins/globalMixin.js'

Vue.mixin(globalMixin)
```
When I attempt to do this in Nuxt 3, I get the error 'Cannot read property mixin of undefined'.

After some experimentation I was able to change my import so that it worked with es6 by doing `import * as Vue from 'vue'`. This solved the issue above, but when I tried to do the same command above `Vue.mixin(globalMixin)` I got a new error '__vite_ssr_import_0__.mixin is not a function'. Can anybody help me add a global mixin or is there another more effective way to make a group of functions available globally?

**Comment by TheAlexLichter** [comment, marked as answer]
If you want to use a global mixin, you should register it in a plugin (no matter if you use Nuxt2 or Nuxt3 😋) and not in the layout.

The global `Vue.use` API does not exist in Vue 3 anymore. Instead, you have to get the `app` somehow and call `use` there.

Here is an examplary Nuxt plugin that will apply the global mixin as you desired.

```js
import globalMixin from '~/mixins/globalMixin.js'

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.use(globalMixin)
})

**Reply by RezaErfani67** [reply]
@manniL  how can i use it in component?
every thing i tried give me fault result

**Reply by RezaErfani67** [reply]
my mixin file:
```
export default {

   

    methods:{
          callme(){
            console.log("not running")
          },
    },

    
}
```

my plugin.js:
```
import globalMixin from '~/mixins/global'

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.use(globalMixin)
})
```

**Reply by Krislunde** [reply]
@RezaErfani67 : Please have a look at https://v3.nuxtjs.org/guide/directory-structure/plugins and then use your mixin exactly like a plugin.

**Reply by Daedra22** [reply]
> @RezaErfani67 : Please have a look at https://v3.nuxtjs.org/guide/directory-structure/plugins and then use your mixin exactly like a plugin.

The docs are not clear, especially how to nest a method.
Could you provide an example instead?

**Reply by gavin667** [reply]
If you do something like the below it should work as you intend. I believe the `.use` method is for registering Vue directives which operate slightly different from mixins and as such are registered differently.

```
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.mixin(globalMixin)
})
```

Also just to note but there is a move away from `mixins` entirely and they are [no longer recommended](https://vuejs.org/api/options-composition.html#mixins). Not always an easy feat to dump use of mixins completely if you are migrating from Nuxt/Vue 2 to 3 though! Hopefully the above works. Have fun!

_P.S Here for any more issues you might have!_

**Reply by usaravanan** [reply]
@gavin667 Excellent!!! It works like Charm. I was struggling for half of the day to use mixin in nuxt. Thanks a lot.

**Comment by Trevash** [comment]
Awesome, thank you!

**Comment by Venegrad** [comment]
Not working
